clinica/iotools/converters/ukb_to_bids/ukb_utils.py

```diff
@@ -0,0 +1,344 @@
+                },
+                "SWI.nii.gz": {
+                    "datatype": "swi",
+                    "modality": "swi",
+                    "suffix": "swi",
+                    "sidecars": [],
+                    "task": "",
+                    "dir": "",
+                },
+            }
+        ).apply(pd.Series)
+    )
+    df_clinical = df_clinical.assign(year_of_birth=lambda df: df.year_of_birth_f34_0_0)
+
+    df_clinical = df_clinical.assign(age=lambda df: df.age_at_recruitment_f21022_0_0)
+    df_clinical = df_clinical.assign(
+        age_at_first_session=lambda df: df.age_when_attended_assessment_centre_f21003_2_0
+    )
+    df_clinical["age_at_session"] = df_clinical.apply(
+        lambda df: select_session(df), axis=1
+    )
+    df_clinical = df_clinical.assign(
+        session_month=lambda df: (df.age_at_session - df.age_at_first_session) * 12
+    )
+
+    df_clinical = df_clinical.assign(
+        session=lambda df: df.session_month.map(lambda x: f"ses-M{x:03d}")
+    )
+    df_clinical = df_clinical.assign(
+        bids_filename=lambda df: (
+            "sub-UKB"
+            + df.source_id.astype("str")
+            + "_"
+            + df.session.astype("str")
+            + df.dir.astype("str")
+            + df.task.astype("str")
+            + "_"
+            + df.suffix
+        )
+    )
+    df_clinical = df_clinical.join(
+        df_clinical.sex_f31_0_0.astype("str")
+        .map(
+            {
+                "0": {"sex": "F"},
+                "1": {"sex": "M"},
+            }
+        )
+        .apply(pd.Series)
+    )
+    df_clinical = df_clinical.assign(
+        bids_full_path=lambda df: df.participant_id
+        + "/"
+        + df.session
+        + "/"
+        + df.datatype
+        + "/"
+        + df.bids_filename
+    )
+
+    return df_clinical
+
+
+def dataset_to_bids(df_clinical: DataFrame) -> Tuple[DataFrame, DataFrame, DataFrame]:
+
+    import os
+
+    import pandas as pd
+
+    # open the reference for building the tsvs:
+    path_to_ref_csv = os.path.join(
+        os.path.dirname(os.path.dirname(os.path.dirname(__file__))),
+        "data",
+        "ukb_ref.csv",
+    )
+    df_ref = pd.read_csv(path_to_ref_csv, sep=";")
+
+    df_clinical = df_clinical.set_index(
+        ["participant_id", "session", "modality", "bids_filename"],
+        verify_integrity=True,
+    )
+
+    # Build participants dataframe
+    df_participants = df_clinical.filter(items=list(df_ref["participants"]))
+
+    # Build sessions dataframe
+    df_session = df_clinical.filter(items=list(df_ref["session"]))
+
+    # Build scans dataframe
+    df_scan = df_clinical.filter(items=list(df_ref["scan"]))
+    return df_participants, df_session, df_scan
+
+
+def write_to_tsv(dataframe: DataFrame, buffer: Union[PathLike, BinaryIO]) -> None:
+    # Save dataframe as a BIDS-compliant TSV file.
+    dataframe.to_csv(buffer, sep="\t", na_rep="n/a", date_format="%Y-%m-%d")
+
+
+def write_bids(
```

Comment:
It feels like this function isn't specific to the `ukb_to_bids` converter. It could be worth having it as a utility function and sharing it across converters. WDYT?

Reply:
I'd second that opinion. But for the sake of restricting the scope of this PR, I'd suggest just moving the function to a common utility module and not refactoring the interface with other converters.

With "recent" converters such as `HABS` or `Oassis3` the integration should require minimal effort but might be a bit more costly with other converters. At any rate, best to separate this effort from the core converter.

Reply:
The write_bids function may be a little different for each converter depending on the data we have ! but the write_to_tsv function right above was moved 😄